### PR TITLE
add GRASS

### PR DESCRIPTION
**Pull request by Crghilardi**
This PR adds a `build_tarballs.jl` for the [GRASS](https://grass.osgeo.org/) libraries. 

This will need some work, I only tested `x86_64-linux-gnu`  but I wanted to be able to look at CI. I know there are a few separate configure flags for mac and windows platforms when I run `./configure --help` that I have not added in yet.

This builds all but one of the libraries (`g.proj`) which was returning segmentation faults

```
make[1]: Leaving directory '/workspace/srcdir/grass-7.8.5/general/g.ppmtopng'
make -C g.proj || echo /workspace/srcdir/grass-7.8.5/general/g.proj >> /workspace/srcdir/grass-7.8.5/error.log
make[1]: Entering directory '/workspace/srcdir/grass-7.8.5/general/g.proj'
if [ "/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/bin/g.proj" != "" ] ; then GISRC=/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/demolocation/.grassrc78 GISBASE=/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu PATH="/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/bin:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/bin:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/scripts:$PATH" PYTHONPATH="/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/etc/python:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/gui/wxpython:$PYTHONPATH" LD_LIBRARY_PATH="/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/bin:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/bin:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/scripts:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/lib:/workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/lib:/usr/local/lib64:/usr/local/lib:/usr/lib64:/usr/lib:/lib64:/lib:/usr/lib/csl-musl-x86:/usr/lib/csl-glibc-x86:/usr/lib/csl-musl-i686_64:/usr/lib/csl-glibc-i686_64:/opt/x86_64-linux-musl/x86_64-linux-musl/lib64:/opt/x86_64-linux-musl/x86_64-linux-musl/lib:/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib64:/opt/x86_64-linux-gnu/x86_64-linux-gnu/lib:/workspace/destdir/lib64:/workspace/destdir/lib" LC_ALL=C LANG=C LANGUAGE=C /workspace/srcdir/grass-7.8.5/dist.x86_64-pc-linux-gnu/bin/g.proj --html-description < /dev/null | grep -v '</body>\|</html>' > g.proj.tmp.html ; fi
proj_get_authorities_from_database: Open of /workspace/destdir/share/proj/proj.db failed
Segmentation fault (core dumped)
make[1]: *** [../../include/Make/Html.make:14: g.proj.tmp.html] Error 1FYI 

```

FYI @visr @evetion

### CODE REVIEW

File: G/GRASS/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="GDAL_jll", uuid="a7073274-a066-55f0-b90d-d619367d196c"))
+    Dependency(PackageSpec(name="PROJ_jll", uuid="58948b4f-47e0-5654-a9ad-f609743f8632"))
+    Dependency(PackageSpec(name="Cairo_jll", uuid="83423d85-b0ee-5818-9007-b63ccbeb887a"))
+    Dependency(PackageSpec(name="FreeType2_jll", uuid="d7e528f0-a631-5988-bf34-fe36492bcfd7"))
+    Dependency(PackageSpec(name="LAPACK_jll", uuid="51474c39-65e3-53ba-86ba-03b1b862ec14"))
+    Dependency(PackageSpec(name="FFTW_jll", uuid="f5851436-0d7a-5f13-b9de-f02708fd171a"))
+    Dependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
+    Dependency(PackageSpec(name="Xorg_xorgproto_jll", uuid="c4d99508-4286-5418-9131-c86396af500b"))
```
Comment (giordano) [inline G/GRASS/build_tarballs.jl:530]: This is a build dependency, these are only header files
```suggestion
    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid="c4d99508-4286-5418-9131-c86396af500b"))
```

File: G/GRASS/bundled/patches/remove_gproj.patch [hunk 2]
```diff
@@ -0,0 +1,11 @@
+--- before/general/Makefile
++++ after/general/Makefile
+@@ -19,7 +19,6 @@ SUBDIRS = \
+ 	g.parser \
+ 	g.pnmcomp \
+ 	g.ppmtopng \
+-	g.proj \
++	#g.proj \
```
Comment (giordano) [inline G/GRASS/bundled/patches/remove_gproj.patch:8]:
```suggestion
```

**Comment by giordano** [comment]
For Windows we probably need to do
https://github.com/JuliaPackaging/Yggdrasil/blob/0063d7f62a447270bf31614d6490e4afe3743e9a/O/OpenLDAPClient/build_tarballs.jl#L25-L29
https://github.com/JuliaPackaging/Yggdrasil/blob/0063d7f62a447270bf31614d6490e4afe3743e9a/O/OpenLDAPClient/build_tarballs.jl#L60-L63

**Comment by giordano** [comment]
It doesn't look like this can be cross-compiled, not as it now at least: it's trying to run an application built for the target platform

**Comment by giordano** [comment]
It should be possible to cross-compile by bootstrapping first a native build, some information specific for MinGW:

* https://idea.isnew.info/how-to-cross-compile-grass-gis-for-ms-windows.html
* https://github.com/HuidaeCho/grass-build-scripts

We do something similar for ICU:
https://github.com/JuliaPackaging/Yggdrasil/blob/0063d7f62a447270bf31614d6490e4afe3743e9a/I/ICU/build_tarballs.jl#L25-L47
Not super straightforward, but doable once one figures out all relevant flags to pass to `./configure` (the links above should help)

**Comment by Crghilardi** [comment]
There is an [--with-macosx-archs[=arch list]](https://github.com/OSGeo/grass/blob/master/configure#L24) flag that I have not played with yet, that was going to be my next step. Do you think there is any value in that option?

**Comment by giordano** [comment]
You can use `x86_64`, I'm not sure how they detect the "native" platform

**Comment by giordano** [comment]
Ok, very small step forward: we can build for the host system, which is useful in case we need to do the bootstrap for the other platforms.

I'm reading the instructions in [How to cross-compile GRASS GIS for MS Windows](https://idea.isnew.info/how-to-cross-compile-grass-gis-for-ms-windows.html) but I'm not really sure about what they do differently between the native build and the target build :thinking:

**Comment by Crghilardi** [comment]
Yeah, I don't quite understand it either.If I understand correctly, it's just the docs and some associated python scripts that are the friction point?

I think the other non-mingw [repo](https://github.com/HuidaeCho/grass-build-scripts/blob/master/myconfigure.sh#L74) seems to show something similar to the ICU example you linked to.

**Comment by giordano** [comment]
> Yeah, I don't quite understand it either.If I understand correctly, it's just the docs and some associated python scripts that are the friction point?

That's my understanding as well :disappointed:

**Comment by giordano** [comment]
Ok, now I have the prebuilt package for bootstrap, but I'm quite lost about what exactly we have to do to use it.

**Comment by giordano** [comment]
What do you need exactly from GRASS?  It looks like `make -C libs` builds all the libraries at it should work flawless.  Do you need something else besides the libraries?

**Comment by Crghilardi** [comment]
Having just the libraries should work. I tried figuring out how to leave out the stuff like the GUI pieces for a bit when I first started but never got there.

**Comment by giordano** [comment]
> It looks like `make -C lib` builds all the libraries at it should work flawless.

Ugh, no, only _some_ libraries are built :persevere:

**Comment by giordano** [comment]
I feel like they have a bizarre build system, with the interdependencies not well specified.  As the script is now, we can build most libraries, but some are missing because they depend on other libraries not yet built....

**Comment by Crghilardi** [comment]
Yeah, I don't have any useful insights at this point either. I might just make an account for the [grass-dev](https://lists.osgeo.org/mailman/listinfo/grass-dev) mailing list to ask about cross-compiling specifically. Might make for a good addition to the upstream wiki.

**Comment by giordano** [comment]
Ok, now only `libgrass_lrs`, `libgrass_iortho`, `libgrass_sim` and `libgrass_rli` are missing.

**Comment by giordano** [comment]
So much green now :slightly_smiling_face:  But having to manually specify all those targets in the right order is rather annoying

**Comment by giordano** [comment]
Ok, I think we're finally in a good shape!  All platforms are building without serious errors (there are still some errors thrown, but they don't seem to be fatal).

Are you in a rush with this package?  In the coming weeks I'm going to rebuild some of the dependencies of this package, which may or may not (I'll find it out along the way) introduce some incompatibilities with older versions.  If you aren't in a hurry with this one, I'd suggest to wait a bit until the relevant dependencies are rebuilt so that we'll know whether there are relevant incompatibilities to deal with or not.

**Comment by Crghilardi** [comment]
I am not in any rush with this package, so I will wait and watch. Which deps will be rebuilt?

I have also been reading through some other [mac specific build recipe](https://github.com/qgis/QGIS-Mac-Packager/blob/f12e73d0867c49a9f941127ab902eb7eaf5362c7/qgis_deps/recipes/grass/recipe.sh) and the series of patches for the [conda feedstock](https://github.com/osgeo-forge/grass-feedstock/tree/master/recipe) I found. Have not gotten much out of them myself, but posting here as well JIC.

**Comment by giordano** [comment]
> Which deps will be rebuilt?

The Cairo stack.  Packages relevant here are Cairo_jll, FreeType2_jll and Libiconv_jll, but they seem all good as far as I can see.

File: G/GRASS/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    Dependency(PackageSpec(name="Cairo_jll", uuid="83423d85-b0ee-5818-9007-b63ccbeb887a"))
+    Dependency(PackageSpec(name="FreeType2_jll", uuid="d7e528f0-a631-5988-bf34-fe36492bcfd7"))
+    Dependency(PackageSpec(name="LAPACK_jll", uuid="51474c39-65e3-53ba-86ba-03b1b862ec14"))
+    Dependency(PackageSpec(name="FFTW_jll", uuid="f5851436-0d7a-5f13-b9de-f02708fd171a"))
+    Dependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
+    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid="c4d99508-4286-5418-9131-c86396af500b"))
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"7.1.0")
```
Comment (giordano) [inline G/GRASS/build_tarballs.jl:578]: Are you ok to require Julia v1.6?  It'll make building for experimental platforms in the future easier.

Comment (Crghilardi) [reply G/GRASS/build_tarballs.jl:578]: Yes, Julia 1.6+ sounds perfectly reasonable

Comment (giordano) [reply G/GRASS/build_tarballs.jl:578]:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"7.1.0", julia_compat="1.6")
```

**Pull request merged**